Document user files of nix (#8141)
NixOS/nix: doc/manual/src/command-ref/files/default-nix-expression.md

If the default expression is a directory without a `default.nix` file, then its contents (both files and subdirectories) are loaded as Nix expressions.
The expressions are combined into a single set, each expression under an attribute with the same name as the original file or subdirectory.
Subdirectories without a `default.nix` file are traversed recursively in search of more Nix expressions, but the names of these intermediate directories are not added to the attribute paths of the default Nix expression.

For example, if the default expression contains two files, `foo.nix` and `bar.nix`, then the default Nix expression will be equivalent to

```nix
{
  foo = import ~/.nix-defexpr/foo.nix;
  bar = import ~/.nix-defexpr/bar.nix;
}
```

The file `manifest.nix` is always ignored.

The command [`nix-channel`] places a symlink to the user's current [channels profile](@docroot@/command-ref/files/channels.md) in this directory.
This makes all subscribed channels available as attributes in the default expression.

## User channel link

A symlink that ensures that [`nix-env`] can find your channels.

> - `~/.nix-defexpr/channels -> /nix/var/nix/profiles/per-user/$USER/channels`
> - `$XDG_STATE_HOME/defexpr/channels -> $XDG_STATE_HOME/profiles/channels`

Reviewer: ```suggestion
- `~/.nix-defexpr/channels`
- `$XDG_STATE_HOME/defexpr/channels` if [`use-xdg-base-directories`] is set to `true`.

This symlink points to:

 - `$XDG_STATE_HOME/profiles/channels` for regular users
- `$NIX_STATE_DIR/profiles/per-user/root/channels` for root
```

# Default Nix expression

> The source for the default [Nix expressions](@docroot@/language/index.md) used by [`nix-env`].

Reviewer 2: What is it used for?

- `~/.nix-defexpr`
- `$XDG_STATE_HOME/nix/defexpr` if [`use-xdg-base-directories`] is set to `true`.

> If the default expression is a file, it is loaded as a Nix expression.

Reviewer 2: Handling directories should be grouped together here.

> If the expression is a set, it is used as the default Nix expression.

Reviewer 2: ```suggestion
If the expression is an attribute set, it is used as the default Nix expression.
```

If the expression is a function, an empty set is passed as argument and the return value is used as the default Nix expression.

> If the default expression is a directory containing a `default.nix` file, that file is loaded as in the above paragraph.
> 

Reviewer 2: ```suggestion
```

If the default expression is a directory without a `default.nix` file, then its contents (both files and subdirectories) are loaded as Nix expressions.

> The expressions are combined into a single set, each expression under an attribute with the same name as the original file or subdirectory.

Reviewer 2: ```suggestion
The expressions are combined into a single attribute set, each expression under an attribute with the same name as the original file or subdirectory.
```

  The expressions are combined into a single attribute set, each expression under an attribute with the same name as the original file or subdirectory.
  Subdirectories without a `default.nix` file are traversed recursively in search of more Nix expressions, but the names of these intermediate directories are not added to the attribute paths of the default Nix expression.

Then, the resulting expression is interpreted like this:

- If the expression is an attribute set, it is used as the default Nix expression.
- If the expression is a function, an empty set is passed as argument and the return value is used as the default Nix expression.

For example, if the default expression contains two files, `foo.nix` and `bar.nix`, then the default Nix expression will be equivalent to

```nix
{
  foo = import ~/.nix-defexpr/foo.nix;
  bar = import ~/.nix-defexpr/bar.nix;
}
```

The file `manifest.nix` is always ignored.

The command [`nix-channel`] places a symlink to the user's current [channels profile](@docroot@/command-ref/files/channels.md) in this directory.
This makes all subscribed channels available as attributes in the default expression.

## User channel link

> A symlink that ensures that [`nix-env`] can find your channels.

Reviewer 2: ```suggestion
A symlink that ensures that [`nix-env`] can find your channels:
```
